File: chromadb/utils/rendezvous_hash.py
# An implementation of https://en.wikipedia.org/wiki/Rendezvous_hashing
from typing import Callable, List, cast
import mmh3

Hasher = Callable[[str, str], int]
Member = str
Members = List[str]
Key = str


def assign(key: Key, members: Members, hasher: Hasher) -> Member:
    """Assigns a key to a member using the rendezvous hashing algorithm"""
    if len(members) == 0:
        raise ValueError("Cannot assign key to empty memberlist")
    if len(members) == 1:
        return members[0]
    if key == "":
        raise ValueError("Cannot assign empty key")

    max_score = -1
    max_member = None

    for member in members:
        score = hasher(member, key)
        if score > max_score:
            max_score = score
            max_member = member

    max_member = cast(Member, max_member)
    return max_member


def merge_hashes(x: int, y: int) -> int:
    """murmurhash3 mix 64-bit"""
    acc = x ^ y
    acc ^= acc >> 33
    acc = (
        acc * 0xFF51AFD7ED558CCD
    ) % 2**64  # We need to mod here to prevent python from using arbitrary size int
    acc ^= acc >> 33
    acc = (acc * 0xC4CEB9FE1A85EC53) % 2**64
    acc ^= acc >> 33
    return acc


def murmur3hasher(member: Member, key: Key) -> int:
    """Hashes the key and member using the murmur3 hashing algorithm"""
    member_hash = mmh3.hash64(member, signed=False)[0]
    key_hash = mmh3.hash64(key, signed=False)[0]
    return merge_hashes(member_hash, key_hash)

File: go/coordinator/internal/utils/rendezvous_hash.go
package utils

import (
	"errors"

	"github.com/spaolacci/murmur3"
)

type Hasher = func(member string, key string) uint64
type Member = string
type Members = []Member
type Key = string

// assign assigns a key to a member using the rendezvous hashing algorithm.
func Assign(key Key, members Members, hasher Hasher) (Member, error) {
	if len(members) == 0 {
		return "", errors.New("cannot assign key to empty member list")
	}
	if len(members) == 1 {
		return members[0], nil
	}
	if key == "" {
		return "", errors.New("cannot assign empty key")
	}

	maxScore := uint64(0)
	var maxMember Member

	for _, member := range members {
		score := hasher(string(member), string(key))
		if score > maxScore {
			maxScore = score
			maxMember = member
		}
	}

	return maxMember, nil
}

func mergeHashes(a uint64, b uint64) uint64 {
	acc := a ^ b
	acc ^= acc >> 33
	acc *= 0xff51afd7ed558ccd
	acc ^= acc >> 33
	acc *= 0xc4ceb9fe1a85ec53
	acc ^= acc >> 33
	return acc
}

// NOTE: The python implementation of murmur3 may differ from the golang implementation.
// For now, this is fine since go and python don't need to agree on any hashing schemes
// but if we ever need to agree on a hashing scheme, we should verify that the implementations
// are the same.
func Murmur3Hasher(member string, key string) uint64 {
	hasher := murmur3.New64()
	hasher.Write([]byte(member))
	memberHash := hasher.Sum64()
	hasher.Reset()
	hasher.Write([]byte(key))
	keyHash := hasher.Sum64()
	return mergeHashes(memberHash, keyHash)
}
